pkg/policies/controlplane/components/differentiator.go

```diff
@@ -0,0 +1,147 @@
+package components
+
+import (
+	"time"
+
+	"go.uber.org/fx"
+
+	policylangv1 "github.com/fluxninja/aperture/api/gen/proto/go/aperture/policy/language/v1"
+	"github.com/fluxninja/aperture/pkg/config"
+	"github.com/fluxninja/aperture/pkg/notifiers"
+	"github.com/fluxninja/aperture/pkg/policies/controlplane/iface"
+	"github.com/fluxninja/aperture/pkg/policies/controlplane/runtime"
+)
+
+// Differentiator is a component that calculates rate of change per tick.
+type Differentiator struct {
+	window time.Duration
+	// readings are saved in ring buffer
+	readings  []point
+	oldestIdx int
+	newestIdx int
+	// capacity is calculated from window duration divided by tick interval
+	capacity    int
+	initialized bool
+}
+
+type point struct {
+	reading   runtime.Reading
+	timestamp time.Time
+}
+
+// Name implements runtime.Component.
+func (*Differentiator) Name() string { return "Differentiator" }
+
+// Type implements runtime.Component.
+func (*Differentiator) Type() runtime.ComponentType { return runtime.ComponentTypeSignalProcessor }
+
+// NewDifferentiator creates a differentiator component.
+func NewDifferentiator(diffProto *policylangv1.Differentiator) runtime.Component {
+	diff := &Differentiator{
+		window:      diffProto.Window.AsDuration(),
+		oldestIdx:   0,
+		newestIdx:   0,
+		capacity:    0,
+		initialized: false,
+	}
+	return diff
+}
+
+// NewDifferentiatorAndOptions creates a differentiator component and its fx options.
+func NewDifferentiatorAndOptions(diffProto *policylangv1.Differentiator, _ int, _ iface.Policy) (runtime.Component, fx.Option, error) {
+	return NewDifferentiator(diffProto), fx.Options(), nil
+}
+
+// Execute implements runtime.Component.Execute.
+func (d *Differentiator) Execute(inPortReadings runtime.PortToValue, tickInfo runtime.TickInfo) (runtime.PortToValue, error) {
+	if !d.initialized {
+		d.init(tickInfo)
+	}
+
+	inputVal := inPortReadings.ReadSingleValuePort("input")
+	outputVal := runtime.NewReading(0)
+
+	// add input to readings array
+	d.readings[d.newestIdx] = point{reading: inputVal, timestamp: tickInfo.Timestamp()}
```

Comment:
Maybe you forgot to initialize `newestIdx` and `oldestIdx` based on capacity? They would always be zero

Could have used a deque like structure for a simpler implementation?